Instagram Gallery Block: Secondary users handling

Copons opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Part of #15276

~Requires https://github.com/Automattic/wp-calypso/pull/41334~
~Requires D42269-code~

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
To connect an Instagram Gallery block to Instagram, users need to be linked to WPCOM.

If a user is not linked to WPCOM:
- When adding a new block, disable the "Connect to Instagram" button, and show a notice prompting to link their account to WPCOM first.
- When editing an existing connected block, hide the "Disconnect from Instagram" button in the sidebar.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Connect a site to WPCOM, and test with two users (try with an author if possible).
* Main user (the owner of the WPCOM connection):
  - Insert an Instagram Gallery block.
  - Make sure you can connect it to Instagram without any notices.
  - Make sure the "Disconnect" link in the block sidebar is visible.
  - Save the post.
* Secondary user not linked to WPCOM:
  - Reload the post and select the connected Instagram Gallery block.
  - Make sure the "Disconnect" link in the block sidebar is hidden.
  - Insert a new Instagram Gallery block.
  - Make sure the "Connect to Instagram" button is disabled, and you can see a notice prompting to link the account to WPCOM.
    - Click the link. It will open a WPCOM sign up form.
    - Log in with a secondary WPCOM account, and authorize it (it will link the JP user with the WPCOM one).
    - Click the "return to site" button.
    - Make sure it returns to the post editor.
  - Make sure you are now able to connect it to Instagram.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A

Copons committed:
Add a notice for disconnected users

jeherve commented on extensions/blocks/instagram-gallery/edit.js:
```diff
@@ -146,11 +152,27 @@ const InstagramGalleryEdit = props => {
+							'jetpack'
+						) }
+						<br />
+						<Button href={ addQueryArgs( 'admin.php', { page: 'jetpack#/dashboard' } ) } isLink>
```
Instead of linking to the dashboard, maybe we could just add the connect button right there, doing something like this?
https://github.com/Automattic/jetpack/blob/a5ec8c3d067113871b59c829a99df492014781ee/_inc/lib/admin-pages/class.jetpack-react-page.php#L246-L248

Copons replied:
I realize now the "Next steps" section in the PR description is very long. 😝 

tl;dr
I've tried with `build_connect_url` directly: it opens a WPCOM login page, which eventually is supposed to redirect back to the site (redirect URL is the second parameter of the function).
Unfortunately this is not necessarily the case.

<img width="300" alt="Screenshot 2020-04-16 at 18 36 45" src="https://user-images.githubusercontent.com/2070010/79488262-5f4e7c80-8011-11ea-834d-44361d292a0f.png">

There are two "Back" links:
- The secondary, below the form, respects whatever redirect URL I use in `build_connect_url`. 👍 
- The primary one, the big green button, instead always returns to the Calypso plans page. 👎 

In the PR description you can find examples of what I've tried.
I'm not sure if I'm doing something wrong, or if there's something wrong with the login page instead. 🤔

jeherve replied:
hm sorry, I skipped over that part of the description :\

jeherve replied:
> The primary one, the big green button, instead always returns to the Calypso plans page. 👎

That's most likely because of https://github.com/Automattic/wp-calypso/issues/40156

That flow could use some love in general:
https://github.com/Automattic/wp-calypso/issues/24552
https://github.com/Automattic/wp-calypso/issues/8642

Copons replied:
Ah! Thanks for fetching those out!
I've been juggling a few things and my brain cogs got stuck at some point trying to investigate stuff across one too many repos.

I'll see if there's an obvious quick win in that page.

jeherve commented on extensions/blocks/instagram-gallery/edit.js:
```diff
@@ -146,11 +152,27 @@ const InstagramGalleryEdit = props => {
 							: __( 'Connect your Instagram account', 'jetpack' ) }
 					</Button>
+
+					<Notice isDismissible={ false } status="info">
+						{ __(
+							'To connect your Instagram account, you need to link your account to WordPress.com.',
+							'jetpack'
+						) }
+						<br />
+						<Button href={ addQueryArgs( 'admin.php', { page: 'jetpack#/dashboard' } ) } isLink>
+							{ __( 'Open the Jetpack dashboard', 'jetpack' ) }
+						</Button>
+					</Notice>
```
Do we want to only display the notice when you're not connected to WordPress.com yet?

Copons replied:
Yup!
Sorry about this, I realized way too late that the `{ isCurrentUserConnected && <Notice> }` got left out from the commit, and remained in the uncommitted follow-ups!

jeherve reviewed (changes requested):
This may not be linked to this PR but I thought I would mention it anyway. I can't seem to be able to connect to my Instagram account anymore, and I get the following notices when I click the button to connect:

```
[17-Apr-2020 15:06:57 UTC] PHP Notice:  Undefined property: stdClass::$instagram in /wp-content/plugins/jetpack-dev/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-instagram.php on line 102
[17-Apr-2020 15:06:57 UTC] PHP Notice:  Trying to get property 'connect_URL' of non-object in wp-content/plugins/jetpack-dev/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-instagram.php on line 102
```

![image](https://user-images.githubusercontent.com/426388/79584087-3213d400-80ce-11ea-954c-24de456e1072.png)

Copons commented:
> This may not be linked to this PR but I thought I would mention it anyway. I can't seem to be able to connect to my Instagram account anymore, and I get the following notices when I click the button to connect:

@jeherve It might be partly related.
I've had a similar (very likely the same) issue and recored it in https://github.com/Automattic/jetpack/issues/15276 (see "The API might return null" section).
In my case, it only happened for my secondary account though.
My primary (owner of the JP connection) was and still is able to connect to Instagram via WPCOM.

scruffian commented:
I don't understand why we need this; I thought people couldn't use blocks without a connected account...

jeherve commented:
> I thought people couldn't use blocks without a connected account...

You can. There are different levels of "disconnected" at play here:

- Your site could be using Jetpack's Development mode; in this case you would only get access to a subset of Jetpack blocks, that do not require a connection to WordPress.com. The Calendly or Eventbrite blocks are good examples.
- Another admin could have connected Jetpack to their WordPress.com account; the site would be connected to WordPress.com, but your own account would not be linked. That difference between user link to wpcom and site link to wpcom is important for features that are tied to your own person, like Publicize or Instagram Galllery. In this case, each user has to link their own WordPress.com account to their own account on the site, so they can then link to their own Instagram account.

Copons committed:
Add a proper WPCOM connection flow

Copons commented:
@jeherve I've located where the authorization flow messes up the redirection and pushed a workaround in Calypso (https://github.com/Automattic/wp-calypso/pull/41334).

This is the first of many functions that determine where to redirect after the authorization flow:

https://github.com/Automattic/wp-calypso/blob/75839ebd037c9e33906accc2d04022baa14011ef/client/jetpack-connect/authorize.js#L199-L227

As it seems, our case is not considered by the various conditions, so in my Calypso PR I've added a new one: if the connection URL has a `from=jetpack-block-editor` parameter, it redirects to whatever `redirect_uri` is provided, which in our case is the current post editor.

This workaround is consistent with other established patterns (see the `isFromJpo()` and `isWooRedirect()` methods, that do the same thing with different `from` values).

Unfortunately, I can't seem to test this on Calypso Dev.
Since there are a few redirections in place, my test consisted in letting JP create the connection URL to `https://wordpress.com` then, once on the Jetpack Connect page, replace the domain with `http://calypso.localhost:3000` and reload.

For some reasons, it always show me the sign up form, even if I'm already logged in to WPCOM (with a secondary account, non linked to JP).
If I try to log in, or even to create a new account, the form eventually always reverts to the sign up form.

Since the redirection happens on authorization complete, I can't effectively test this flow.
Though, the Calypso change is harmless enough, and I think we could just merge it and test on Calypso Staging instead.

Copons committed:
Add a notice for non-admin users

Copons committed:
Instagram connection URL fetch error handling

Copons committed:
Remove unnecessary capability check

Copons commented:
I've updated the PR description with all the new notes and testing instructions!

Copons committed:
Move the WPCOM connection to an external hook

Also force a save if the post is an auto draft,
in order to have the post ID in the redirection URL.

Copons commented:
In 0efaf00 I've addressed an edge case reported by @scruffian: since the redirection URL is created with a simple `window.location.href`, if the post is not saved when the Instagram Gallery block is inserted, the WPCOM connection would end up redirecting back to a new post.

The fix checks if the post is an `auto-draft`, and in that case the post is saved before the redirection URL is created.

As an aside, since the logic was becoming quite complicated, I've moved it in an external `useConnectWpcom` hook.
It might even turn out useful for other blocks in the future, who knows!

scruffian commented:
When I disconnect my user account from a WPCOM, I don't see the notice in the Editor, so I can't reconnect it - I just get an error in the editor.

scruffian commented:
I'm also getting an error when I try to connect to the `wpcom/v2/instagram/connect-url` endpoint, which means I can't make connections with my normal user either...

scruffian commented:
I tested this again and it's all working, sorry about that; I must have had something wrong in my dev setup.

I noticed a couple of things but they shouldn't stop us merging this.

1. The padding around this notice looks off. It should probably be above the button too:
<img width="752" alt="Screenshot 2020-04-24 at 14 38 06" src="https://user-images.githubusercontent.com/275961/80219370-3d20b400-863a-11ea-8c52-38a07dbda439.png">

2. I see this notice on Calypso, which doesn't make sense since this is a new account:
<img width="543" alt="Screenshot 2020-04-24 at 14 18 41" src="https://user-images.githubusercontent.com/275961/80219355-3a25c380-863a-11ea-80ae-65113fdae370.png">
The flow still works though

scruffian reviewed (dismissed):
Code looks good and works as expected 🚢

Copons committed:
Update the placeholder for users not connected to WPCOM

Copons commented:
@scruffian I've addressed the UI inconsistency with the Notice in d3d5c5b.
Now it's much simpler: instead of showing all the things, the steps are cleanly separated:

| Not connected to WPCOM | Connected to WPCOM but not to Instagram |
| - | - |
| <img width="534" alt="Screenshot 2020-04-24 at 15 20 49" src="https://user-images.githubusercontent.com/2070010/80223333-0ea5d780-8640-11ea-855d-3fcbecf7aaca.png"> | <img width="538" alt="Screenshot 2020-04-24 at 15 21 12" src="https://user-images.githubusercontent.com/2070010/80223358-18c7d600-8640-11ea-8853-61a84a799277.png"> |

I think we might want to refine the copy though, because I've really overused the word "account". 😅 

@davemart-in mind offering some help here?
We basically have 2 disconnected states in the block, that need to be performed in sequence: first you need to connect your site account to WPCOM (this only happens on JP), then you need to connect your WPCOM account to Instagram.

In both cases, in the UI we have room for a button and an explanation.
As you can see in the screens above, in the second case there's only the button, but we might want to explain it a bit more.

davemart-in commented:
Thanks for the ping @Copons!

Here's what I recommend:

**Instagram Gallery**
First, you'll need to connect to WordPress.com.
[ Connect to WordPress.com ]

***Instagram Gallery**
Connect to Instagram to start sharing your images.
[ Connect to Instagram ]

Copons committed:
Update placeholder copy

jeherve commented on extensions/blocks/instagram-gallery/constants.js:
```diff
@@ -0,0 +1,11 @@
+ */
+import { get } from 'lodash';
+
+export const IS_CURRENT_USER_CONNECTED_TO_WPCOM = get( window.Jetpack_Editor_Initial_State, [
```
Not a blocker for now, but maybe we can move this to a shared function as I imagine other blocks may need this in the future?

Copons replied:
This is just basically getting a global variable, not sure it's worth as a shared utility 😅

jeherve replied:
True, but this may be duplicated in quite a few places in the future. Still, I suppose we can wait until this actually happens :)

jeherve commented on extensions/blocks/instagram-gallery/use-connect-wpcom.js:
```diff
@@ -0,0 +1,53 @@
+/**
```
Maybe this would also be a useful shared function in the future?

jeherve reviewed (dismissed):
This should be good to merge. This flow is really nicely done, nice work! I left 2 comments, but this can be addressed in a future PR.

jeherve reviewed (approved):
One more approval after one more rebase :)

pento reviewed (approved):
🚢 Looks good to me!
